### PR TITLE
Link & SLSA attestor

### PR DESCRIPTION
**Pull request by jkjell**
Add a new link attestor and the ability to return multiple signed attestation with `RunWithExports` function.

### CODE REVIEW

File: attestation/link/link.go
Comment (ChaosInTheCRD) [inline attestation/link/link.go:103]: So this is less a comment specific to this code, but more something i'm wondering more generally:

> Should an attestor be able to inspect the metadata generated by other attestors? Even if the attestor is native to the tool (not a custom attestor like I have proposed elsehwere), is it _safe_ to attestors to basically interact with each others data?

Comment (ChaosInTheCRD) [inline attestation/link/link.go:103]: Am I right in saying that this *must* run last? I know we're running this as a "PostProduct", but could we run the risk of not catching other "PostProduct" attestors? Maybe this functionality would be better served as a flag that generates a link attestation rather than an attestor type?

Comment (jkjell) [reply attestation/link/link.go:103]: I was thinking that a new lifecycle type of "composite" that runs after all the other attestors. The composite attestor could be given a copy of the data instead of direct references. That wouldn't allow someone to "accidentally" modify the results of any other attestor.

Comment (jkjell) [reply attestation/link/link.go:103]: https://github.com/in-toto/go-witness/pull/149/#discussion_r1478435889

File: run.go
Comment (ChaosInTheCRD) [inline run.go:77]: Maybe I am not seeing the link attestor the right way, but to me turning `RunResult` into a slice of results is a bit confusing. It implies in my head that `Run` has been executed more than once. Instead I think I am right in saying that the link predicate is more of a way of packaging up references to the attestations? Given this, wouldn't it make more sense to return the link attestor as a field in the `RunResult` struct?

File: run.go
Comment (ChaosInTheCRD) [inline run.go:155]: I think maybe keeping `signCollection` as is could be the right way here, and then maybe break out the contents to another shared function? 🤷 at the same time this might just be me getting the heebie jeebies by the fact that it still says `collection` for the input var. If this was change to something like `predicate`, maybe it would make more sense?

File: attestation/link/link.go
Comment (ChaosInTheCRD) [reply attestation/link/link.go:103]: Yeah I definitely think something that handled stuff **after** the attestors run would be the way to go here. It also keeps things decoupled as you say, to in the future completely lock attestors out from seeing each others metadata.

**Comment by ChaosInTheCRD** [comment]
some opinionated thoughts from me, one little style thing that I think could be a tiny bit neater, but on the whole I ike the principle of the link connector and providing this way of exposing attestations. Code is also really neat and tidy 😄.

File: attestation/link/link.go
Comment (ChaosInTheCRD) [reply attestation/link/link.go:103]: Was this meant to another PR? this is just taking me to the top of this PR 😃 .

File: run.go
Comment (jkjell) [reply run.go:77]: I have no strong opinions about the naming and structure of the result. The important part is this is the boundary for signing envelopes. I'm actually not sure why RunResult has the `attestation.Collection` either. 🤔

File: attestation/link/link.go
Comment (colek42) [reply attestation/link/link.go:103]: We need to inspect to output from other attestors.  For example in the OCI attestor we query the product attestor for candidates.  It is the reason why we needed to create the lifecycle

Comment (colek42) [reply attestation/link/link.go:103]: would it make sense for the link attester to be post product?  I'm concerned about adding more steps to the lifecycle, and making the user define the order would not be ideal

Comment (jkjell) [reply attestation/link/link.go:103]: > would it make sense for the link attester to be post product
We actually need it to be the last post product. Example would be to add a subject for an OCI image. We have to run after the OCI post product attestor.

File: run.go
Comment (jkjell) [reply run.go:155]: Yeah, renaming to `predicate` makes sense.

File: attestation/link/link.go
Comment (ChaosInTheCRD) [reply attestation/link/link.go:103]: Seems it might be sensible to (at least for now) ensure that this attestor is handled last. I am also guessing that this sort of behaviour would need to apply to the SLSA attestor that sits in this PR. 

We should probably think more about the seemingly increasing number of attestors that will not gather information themselves but instead rely on evidence generated by others. Do we have an existing issue for this @jkjell @mikhailswift ?

Comment (ChaosInTheCRD) [reply attestation/link/link.go:103]: created one https://github.com/in-toto/witness/issues/422. Don't think we need to implement this in order to have this stuff merged though.

File: attestation/slsa/slsa.go
Comment (ChaosInTheCRD) [inline attestation/slsa/slsa.go:112]: @jkjell I think i'm right in saying that we need to add extra custom logic so that we can add this metadata depending on where / how witness was invocated? Maybe something that we might need to think about.

The way I see it, we should (as a first pass) look for attestors such as github, gitlab, google cloud build? If those attestors have run and are populated, we can use those to populate these fields.

File: attestation/slsa/slsa.go
Comment (ChaosInTheCRD) [inline attestation/slsa/slsa.go:115]:
```suggestion
	internalParameters := make(map[string]interface{})
```

File: attestation/slsa/slsa.go
Comment (ChaosInTheCRD) [inline attestation/slsa/slsa.go:164]:
```suggestion
			internalParameters["env"] = pbEnvs
```

File: attestation/slsa/slsa.go
Comment (ChaosInTheCRD) [inline attestation/slsa/slsa.go:172]:
```suggestion
	p.PbProvenance.BuildDefinition.InternalParameters, err = structpb.NewStruct(internalParameters)
```

Comment (ChaosInTheCRD) [reply attestation/slsa/slsa.go:172]: Could we put this in the environment attestor case statement?

File: run.go
Comment (ChaosInTheCRD) [reply run.go:78]: Yeah I was just looking at this. The `attestation.Collection` can be extracted from the `dsse.Envelope`, we could have a function attached to the RunResult type.

File: attestation/slsa/slsa.go
Comment (jkjell) [reply attestation/slsa/slsa.go:172]: There could be a few other internalParameters to add from the other attestors.

File: attestation/slsa/slsa.go
Comment (jkjell) [reply attestation/slsa/slsa.go:112]: Yep, GitHub and GitLab are the two we'll implement for sure. We'll have to come up with a reasonable name for a default or "undetected".

File: attestation/slsa/slsa.go
Comment (ChaosInTheCRD) [inline attestation/slsa/slsa.go:205]: hey @jkjell ! I pulled this down to see if I can make some QOL improvements, and I noticed that I was getting an `assignment to entry in nil map` error. Is this code running for you? I had to modify the code to initialise `p.products` in order to get it to work :)

Comment (ChaosInTheCRD) [reply attestation/slsa/slsa.go:205]: also would it make sense to leverage ctx.Products() here?

File: attestation/slsa/slsa.go
Comment (ChaosInTheCRD) [inline attestation/slsa/slsa.go:208]: - I see that we have the ProductAttestor interface https://github.com/in-toto/go-witness/blob/74a58da23ce6d08323d763807748d7584bad8ade/attestation/product/product.go#L50
- we cast OCI attestor to this interface if it is found:
```
		case oci.Name:
			maps.Copy(p.products, attestor.Attestor.(product.ProductAttestor).Products())
		}
```
- However, from what I can see the Products() function isn't implemented for this attestor :thinking_face:? https://github.com/in-toto/go-witness/blob/link-attestor/attestation/oci/oci.go

Comment (jkjell) [reply attestation/slsa/slsa.go:218]: Yeah, so this actually makes sense now that I think about it. The OCI attestor doesn't actually create any files but, it does create additional subjects for the OCI references it detects. I'll need to think through what subjects we actually want to add to the SLSA provenance attestation. 🤔

File: run.go
Comment (mikhailswift) [reply run.go:75]: I'm operating off of memory here -- but the `dsse.Envelope` stores the Collection as json bytes. Having the Collection as part of the RunResult was to allow users of the library to not have to re-parse the bytes to get the thing we just made for them.

**Comment by mikhailswift** [comment]
Did a quick read through of the code and this all looks reasonable. Will pull and play with it tomorrow.

**Review by ChaosInTheCRD — APPROVED** [review]

**Comment by ChaosInTheCRD** [comment]
provisionally approving, as long as we can resolve https://github.com/in-toto/go-witness/pull/149#discussion_r1552115108

**Comment by ChaosInTheCRD** [comment]
Just noticed as well, we need to fix DCO before merge 😄

File: attestation/slsa/slsa.go
Comment (jkjell) [reply attestation/slsa/slsa.go:222]: So, for now, we're just adding all of the subjects. When we make future improvements to the OCI attestor, we should revisit this.

**Review by ChaosInTheCRD — APPROVED** [review]

**Pull request merged**
